File: 3405-count-the-number-of-special-characters-ii/count-the-number-of-special-characters-ii.py
class Solution:
    def numberOfSpecialChars(self, word: str) -> int:
        lower = {}
        upper = {}

        for i, ch in enumerate(word):
            if ch.islower():
                lower[ch] = i
            else:
                if ch.lower() not in upper:
                    upper[ch.lower()] = i

        count = 0

        for ch in lower:
            if ch in upper and lower[ch] < upper[ch]:
                count += 1

        return count

File: 3405-count-the-number-of-special-characters-ii/count-the-number-of-special-characters-ii.cs
public class Solution {
    public int NumberOfSpecialChars(string word) {
        int n = word.Length;
        Dictionary<char, int> lastLower = new Dictionary<char, int>();
        Dictionary<char, int> firstUpper = new Dictionary<char, int>();

        for (int i = 0; i < n; i++) {
            char ch = word[i];

            if (char.IsLower(ch)) {
                lastLower[ch] = i;
            }
            else {
                char lower = char.ToLower(ch);

                if (!firstUpper.ContainsKey(lower)) {
                    firstUpper[lower] = i;
                }
            }
        }

        int count = 0;

        foreach (var kv in lastLower) {
            char ch = kv.Key;

            if (firstUpper.ContainsKey(ch) && lastLower[ch] < firstUpper[ch]) {
                count++;
            }
        }

        return count;
    }
}